### PR TITLE
Add UFS root before the Alluxio file path

### PR DESCRIPTION
**Pull request by huanghua78**
Signed-off-by: Huang Hua <huanghua78@msn.com>

### What changes are proposed in this pull request?

In Dora client, we need to join the UFS root with the Alluxio file path before calling UfsBaseFileSystem operations.

### Why are the changes needed?

All input FilePath strings from client are Alluxio based file path. So, the root path of the UFS should be added before the Alluxio file path and be passed to UfsBaseFileSystem on client side. The file path returned from UfsBaseFileSystem.getStatus() and UfsBaseFileSystem.listStatus() also need to be stripped with its UFS root.

When the FileStatus is sent back from DoraWorker, the file path in FileInfo needs similar transformation.

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java [hunk 1]
```diff
@@ -212,17 +212,23 @@ public URIStatus getStatus(AlluxioURI path) {
 
   @Override
   public URIStatus getStatus(AlluxioURI path, final GetStatusPOptions options) {
-    return callWithReturn(() -> transformStatus(mUfs.get().getStatus(path.getPath())));
+    String ufsFullPath = PathUtils.concatPath(mRootUFS,
+        CommonUtils.stripPrefixIfPresent(path.getPath(), mRootUFS.getPath()));
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: Should we first test if the `path` URI sent by client is within the same root UFS of this `UfsBaseFileSystem`? I don't think it makes sense to try to access something that is already known to be outside of the root UFS. 

If `mRootUFS` is `hdfs://namenode/` and `path` is `s3://bucket/file`, it makes no sense to try to get the status of `hdfs://namenode/file`. Also, what this code does here is creating something like `hdfs://namenode/s3://bucket/file` which is invalid URI.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 2]
```diff
@@ -109,14 +114,29 @@ public StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> respon
...
+      try {
+        String[] paths = PathUtils.getPathComponents(alluxioFilePath);
+        if (paths.length > 0) {
+          fn = paths[paths.length - 1];
+        } else {
+          fn = alluxioFilePath;
+        }
+      } catch (InvalidPathException e) {
+        fn = alluxioFilePath;
+      }
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:131]: We have `new AlluxioURI(alluxioFilePath).getName()` for this.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/common/src/main/java/alluxio/underfs/UfsStatus.java [hunk 3]
```diff
@@ -55,7 +55,7 @@ protected UfsStatus(String name, boolean isDirectory, String owner, String group
     mOwner = owner;
     mGroup = group;
     mMode = mode;
-    mLastModifiedTimeMs = lastModifiedTimeMs;
+    mLastModifiedTimeMs = lastModifiedTimeMs == null ? Long.valueOf(0) : lastModifiedTimeMs;
```
[the remark this answers is not in the document]
Comment (huanghua78) [reply core/common/src/main/java/alluxio/underfs/UfsStatus.java:58]: If the return value of `null` is required by other components, I can handle this situation in another way.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: In Dora, we will only have a single UFS defined by:
`alluxio.dora.client.ufs.root`
We will not have multiple mount tables.

So, the `path` must be the Alluxio internal path.
It will not, should not contain any Provider (s3://, pdfs://, ...) information.

Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: I know, but `UfsBaseFileSystem` is not only meant for Dora. It should be robust enough to handle that even outside the context of Dora.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 2 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java]: Fixed

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: The `path` is Alluxio Internal Path.
It should be joined with the "ufs root" before passing to ufs.
Otherwise the UFS will not recognize this path, and treat it as invalid.

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java [hunk 4]
```diff
@@ -415,7 +415,8 @@ private void openBlock(BlockReadRequestContext context)
         return;
       }
       BlockReadRequest request = context.getRequest();
-      BlockReader reader = mWorker.createFileReader(request.getOpenUfsBlockOptions().getUfsPath(),
+      BlockReader reader = mWorker.createFileReader(
+          String.valueOf(request.getOpenUfsBlockOptions().getUfsPath().hashCode()),
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: why change the file id from a path string to its hashcode? Is the hashcode space large enough to avoid hash collisions?

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: Yes, exactly, so we should do a precondition check and reject anything that is neither an Alluxio path (with no explicit scheme and authority), nor starts with the root UFS URI. Simply calling `CommonUtils.stripPrefixIfPresent(path.getPath(), mRootUFS.getPath())` cannot do this.
For reference, see [`alluxio.client.file.BaseFileSystem#checkUri`](https://github.com/Alluxio/alluxio/blob/1c781f7de19fbcdeddb1d15c8866bb0735f4c520/core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java#L523)

File: core/common/src/main/java/alluxio/underfs/UfsStatus.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbw9580) [reply core/common/src/main/java/alluxio/underfs/UfsStatus.java]: I prefer to using a `Optional<Long>` which better signifies the fact that the last modified time is not always available.

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java (hunk 4 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: 1. The client side uses the hashcode. So, this is to use the same.
2. The original code is purely implemented for Dora. I think this was a typo.  We need an ID here, not a full path file name. This ID is the file ID. 
3. If we use ufspath as ID, this would cause: (a) if the ufs path is s3a://some_bucket/some_file, the final directory would be: /tmp/ramdisk/LOCAL/s3a:/some_bucket/some_file
Please note, the `s3a://` will be split to `s3:/`, and it can not be re-constructed back.
(b) The PagedStoreDir requires the ID to be a long integer. Otherwise, it will report error.

Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: >  Is the hashcode space large enough to avoid hash collisions?

This is a BIG technical debt!!
We need to take actions to use 64bit integer or even 128bit integer or even higher number of integer as the FileID.
This must be done as soon as possible.
@beinan

File: core/common/src/main/java/alluxio/underfs/UfsStatus.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (huanghua78) [reply core/common/src/main/java/alluxio/underfs/UfsStatus.java]: I understand that in Java world, an Optional<Long> can be used to differentiate the actual value and the absent value.
In most of the other cases, we can just use Long.valueof(0) as the absent/default value.

But anyway, this needs a separate task to fix it.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: `alluxio.client.file.BaseFileSystem#checkUri` is in another class, and we can not call it here.

The input, `ufsStatus` is not from user input. It is from getStatus() and listStatus() from this class.
The format of this input should be considered as valid?

We can add more checking here to validate the path scheme if needed.
@LuQQiu

Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: Do you like the following patch, @dbw9580 ?

```
diff --git a/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java b/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
index 7bf66e26d6..ff322afaed 100644
--- a/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
+++ b/core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java
@@ -12,6 +12,7 @@
 package alluxio.client.file.ufs;
 
 import alluxio.AlluxioURI;
+import alluxio.client.file.BaseFileSystem;
 import alluxio.client.file.FileInStream;
 import alluxio.client.file.FileOutStream;
 import alluxio.client.file.FileSystem;
@@ -80,7 +81,7 @@ import javax.annotation.concurrent.ThreadSafe;
  * UFS. No Alluxio server is involved.
  */
 @ThreadSafe
-public class UfsBaseFileSystem implements FileSystem {
+public class UfsBaseFileSystem extends BaseFileSystem implements FileSystem {
   private static final Logger LOG = LoggerFactory.getLogger(UfsBaseFileSystem.class);
 
   /** Used to manage closeable resources. */
@@ -97,6 +98,7 @@ public class UfsBaseFileSystem implements FileSystem {
    * @param options the ufs file system options
    */
   public UfsBaseFileSystem(FileSystemContext fsContext, UfsFileSystemOptions options) {
+    super(fsContext);
     Preconditions.checkNotNull(fsContext);
     Preconditions.checkNotNull(options);
     mFsContext = fsContext;
@@ -408,6 +410,9 @@ public class UfsBaseFileSystem implements FileSystem {
    */
   private URIStatus transformStatus(UfsStatus ufsStatus) {
     String path = CommonUtils.stripPrefixIfPresent(ufsStatus.getName(), mRootUFS.toString());
+    // adding the leading "/" to the beginning of this path.
+    path = PathUtils.concatPath("/", path);
+    super.checkUri(new AlluxioURI(path));
     AlluxioURI ufsUri = new AlluxioURI(PathUtils.concatPath(mRootUFS, path));
     FileInfo info = new FileInfo().setName(ufsUri.getName())
         .setPath(path)
```


we can have ufsBaseFileSystem to extend BaseFileSystem and re-use the checkUri() method.

Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: I'm not sure about extending from `BaseFileSystem`. Can we just copy paste that piece of code for `checkUri`?

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java (hunk 4 above)
Comment (dbw9580) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: >  if the ufs path is s3a://some_bucket/some_file, the final directory would be: /tmp/ramdisk/LOCAL/s3a:/some_bucket/some_file

This is a implementation flaw of `LocalPageStore`. How about using a non-hash codec to transform the path into something that does not resemble a path? e.g. hexadecimal encode that path?

> The PagedStoreDir requires the ID to be a long integer

I think you are referring to `BlockPageId` which was used for blocks that have a numeric ID. It did require that file IDs to be numeric, but this assumption does not hold any more for Dora. Some of the code in `PagedBlockMetaStore` and `PagedBlockStore` needs to be generalized to work with arbitrary file IDs.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (JiamingMai) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: We can add a method like `checkUri(AlluxioURI path)` here to check whether the path string contains schema or not.

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java (hunk 4 above)
Comment (JiamingMai) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: > 1. The client side uses the hashcode. So, this is to use the same.
> 2. The original code is purely implemented for Dora. I think this was a typo.  We need an ID here, not a full path file name. This ID is the file ID.
> 3. If we use ufspath as ID, this would cause: (a) if the ufs path is s3a://some_bucket/some_file, the final directory would be: /tmp/ramdisk/LOCAL/s3a:/some_bucket/some_file
>    Please note, the `s3a://` will be split to `s3:/`, and it can not be re-constructed back.
>    (b) The PagedStoreDir requires the ID to be a long integer. Otherwise, it will report error.

I tested the codes and found that the third problem mentioned above still exists. It is right to avoid regarding UFS path as file id. There must be somewhere that is still using the UFS path as file id.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: Copy & paste is not the right way to __duplicate__ code.
Maybe I can move this function to some common place, e.g. in AlluxioUri.checkUri().
What do you think?

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java (hunk 4 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: we may need a common encode function for the purpose of converting a ufs path to a valid path string.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: But again, as I said, this path is purely internal and returned from listStatus() or getStatus().
It should be a valid ufs path. No need to check it again.

File: core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java (hunk 4 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/FileReadHandler.java:419]: > I tested the codes and found that the third problem mentioned above still exists. It is right to avoid regarding UFS path as file id. There must be somewhere that is still using the UFS path as file id.

Can you please test and tell me where the place is.

File: core/common/src/main/java/alluxio/AlluxioURI.java [hunk 5]
```diff
@@ -499,4 +502,30 @@ public String toString() {
     mUriString = sb.toString();
     return mUriString;
   }
+
+  /**
+   * Computes the hash of this URI, with SHA-256, or MD5, or simple hashCode().
+   *
+   * @return HEX encoded hash string
+   */
+  public String hash() {
```
Comment (JiamingMai) [inline core/common/src/main/java/alluxio/AlluxioURI.java:511]: It would be better to add a unit test method for this `hash()` in `AlluxioURITest`.

**Review by JiamingMai — COMMENTED** [review]
LGTM

File: core/common/src/main/java/alluxio/AlluxioURI.java [hunk 6]
```diff
@@ -499,4 +502,30 @@ public String toString() {
...
+    try {
+      MessageDigest md = MessageDigest.getInstance("SHA-256");
+      md.update(toString().getBytes());
+      return Hex.encodeHexString(md.digest()).toLowerCase();
+    } catch (NoSuchAlgorithmException e) {
+      /* No actions. Continue with other hash method. */
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/AlluxioURI.java:517]: I don't think it's necessary to consider backup algorithms. 1) any decent platform should have implementation for SHA-256, otherwise they cannot support fundamental things like TLS. 2) if on the client side SHA-256 is available but on the server side is not, the hash will be different for the same URI. This will cause page lookup to fail.

We should stick to only one algorithm, and if SHA-256 isn't available, just throw an exception.

File: core/common/src/main/java/alluxio/AlluxioURI.java (hunk 5 above)
Comment (dbw9580) [inline core/common/src/main/java/alluxio/AlluxioURI.java:511]: How about putting this in the `alluxio.underfs.FileId` class (probably need to move that to a different package and module). Maybe add a method `FileId.fromUri(AlluxioURI)`.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: > Can we just copy paste that piece of code for checkUri?

Sorry about not making this clear. Apparently the implementation in `BaseFileSystem` cannot be reused as is, since it is trying to acquire the master authority to check if the authorities match. We don't need that for a `UfsBaseFileSystem` since no master should be involved at all. The implementation of `checkUri` for `UfsBaseFileSystem` should be as simple as

```java
if (!mRootUFS.isAncestorOf(path)) {
  throw new InvalidPathException(...);
}
```

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 7]
```diff
@@ -109,14 +114,20 @@ public StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> respon
...
+      String fn = new AlluxioURI(alluxioFilePath).getName();
+
+      UfsFileStatus status = mUfsFileStatusCache.get(ufsFullPath);
       GetStatusPResponse response = GetStatusPResponse.newBuilder()
           .setFileInfo(
               FileInfo.newBuilder()
-                  .setName(status.getName())
-                  .setPath(ufsFilePath)
-                  .setUfsPath(ufsFilePath)
+                  .setFileId(ufsFullPath.hashCode())
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:127]: need change to the new way of producing a hash

**Comment by jja725** [comment]
@huanghua78 could you elaborate more detail in the description about the intention of the PR and what problem it's trying to resolve? Thanks

File: core/common/src/main/java/alluxio/AlluxioURI.java (hunk 5 above)
Comment (huanghua78) [reply core/common/src/main/java/alluxio/AlluxioURI.java:511]: OK

File: core/common/src/main/java/alluxio/AlluxioURI.java (hunk 6 above)
Comment (huanghua78) [reply core/common/src/main/java/alluxio/AlluxioURI.java:517]: “robustness” is to handle different situations as much as possible.
And "don't" throw exception if itself can handle such error.

"any decent platform", actually, this MessageDigest package is from external package.
We are not sure if it will provide such methods. So we may fall back to use Object.hashCode() if all of these are not available.

File: core/common/src/main/java/alluxio/AlluxioURI.java (hunk 5 above)
Comment (huanghua78) [reply core/common/src/main/java/alluxio/AlluxioURI.java:511]: I think it is better to have it as a public method of AlluxioUri.
1. If FileId.fromUri(AlluxioURI) is needed, it can use the AlluxioUri.hash() directly.
2. This method can also be used for other purpose, e.g. hashing a path to different workers from client.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 7 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:127]: This field requires a "long" type, and we can not use the new SHA256 hash here.

Also, please note, the "FileId" of a FileInfo is a fake/bogus one, and it will not used by client.
On client, I've also made the changes to use new SHA256 hash as its path in local page cache.

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:216]: OK. Let me try.

**Comment by huanghua78** [comment]
> @huanghua78 could you elaborate more detail in the description about the intention of the PR and what problem it's trying to resolve? Thanks


I have create a new PR (https://github.com/Alluxio/alluxio/pull/16847#issuecomment-1423768195) for the PageId hash.
@dbw9580 
@JiamingMai

File: integration/fuse/src/test/java/alluxio/fuse/ufs/stream/InOrOutStreamOutTest.java [hunk 8]
```diff
@@ -51,6 +51,10 @@ protected FuseFileStream createStream(AlluxioURI uri, boolean truncate) {
   @Test(expected = UnimplementedRuntimeException.class)
   public void read() throws Exception {
     AlluxioURI alluxioURI = getTestFileUri();
+    if (alluxioURI.getParent().getPath().length() > 1 ) {
```
Comment (dbw9580) [inline integration/fuse/src/test/java/alluxio/fuse/ufs/stream/InOrOutStreamOutTest.java:54]:
```suggestion
    if (!alluxioURI.getParent().isRoot()) {
```

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 9]
```diff
@@ -176,6 +176,8 @@ private int createOrOpenInternal(String path, FuseFileInfo fi, long mode) {
       return res;
     }
     try {
+      mFileSystem.createDirectory(uri.getParent(),
+          CreateDirectoryPOptions.newBuilder().setRecursive(true).build());
```
Comment (dbw9580) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java]: why is this necessary?

File: integration/fuse/src/test/java/alluxio/fuse/ufs/stream/InOrOutStreamOutTest.java (hunk 8 above)
Comment (huanghua78) [reply integration/fuse/src/test/java/alluxio/fuse/ufs/stream/InOrOutStreamOutTest.java]: Thank you.
Sorry for confusion, but this change is removed.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 9 above)
Comment (huanghua78) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java]: Thank you, but this change is removed.

Actually, because if the dir is a multiple level dir, then create file will fail because its parent dir does not exist.

**Comment by huanghua78** [comment]
Hello @LuQQiu , please review this PR.
Some of the changes are related to UfsBaseFileSystem and some FuseIntegration Testing.
These changes are required because:
The DoraCacheFileSystem expects file path to be Alluxio based file path,
while the UfsBaseFileSystem expects full/absolute path.
We need to do transformation/conversion between them.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 9 above)
Comment (huanghua78) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java]: Let me explain more here, because in latest patch, this is still needed.

In our tests, we may create some files in advance, write some data into these files.
Then tests may read from these files.
These files are created/filled in UFS. In UfsBaseFileSystem, the input path will be treaded absolute/full path in UFS.
But the file path passed to other layer of file system is Alluxio based path.

If the file path is interpreted as Alluxio path, then some parent directories maybe don't exist.
We need to create those directories first.

**Comment by LuQQiu** [comment]
@huanghua78 discussed with beinan that, we want dora a shared cache between different UFSes, so a full ufs path is better to be used throughout full dora from client to worker, so that data corruption issues can be avoided when multiple UFSes with same relative path are used

**Comment by huanghua78** [comment]
> @huanghua78 discussed with beinan that, we want dora a shared cache between different UFSes, so a full ufs path is better to be used throughout full dora from client to worker, so that data corruption issues can be avoided when multiple UFSes with same relative path are used

What is the next step to achieve that goal?

**Comment by dbw9580** [comment]
> so a full ufs path is better to be used throughout full dora from client to worker

Can we somehow encode this requirement as types in code? I'm talking about not using plain `String`s or `AlluxioURI`s, but a dedicated type that no one would be confused about what paths or URIs it contains.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java [hunk 10]
```diff
@@ -176,7 +177,9 @@ private int createOrOpenInternal(String path, FuseFileInfo fi, long mode) {
       return res;
     }
     try {
-      FuseFileStream stream = mStreamFactory.create(uri, fi.flags.get(), mode);
+      FuseFileStream stream = DoraCacheFileSystem.class.isInstance(mFileSystem)
```
Comment (LuQQiu) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:180]: can put in the PathResolver, i can deal with this part in later PR

File: core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java [hunk 11]
```diff
@@ -395,19 +395,19 @@ public void needsSync(AlluxioURI path) throws IOException, AlluxioException {
...
+        .setFileId(ufsUri.toString().hashCode())
         .setUfsPath(ufsUri.toString())
         .setFolder(ufsStatus.isDirectory())
         .setOwner(ufsStatus.getOwner())
         .setGroup(ufsStatus.getGroup())
         .setMode(ufsStatus.getMode())
         .setCompleted(true);
     if (ufsStatus.getLastModifiedTime() != null) {
-      info.setLastModificationTimeMs(info.getLastModificationTimeMs());
+      info.setLastModificationTimeMs(ufsStatus.getLastModifiedTime());
```
Comment (LuQQiu) [inline core/client/fs/src/main/java/alluxio/client/file/ufs/UfsBaseFileSystem.java:410]: Thanks for fixing it!

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java [hunk 12]
```diff
@@ -109,14 +113,19 @@ public StreamObserver<ReadRequest> readBlock(StreamObserver<ReadResponse> respon
...
```
Comment (LuQQiu) [inline core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:125]: Looks like the bug fix is not related to whether we use relative path or absolute path. always using full path can reduce many of the conversion overhead and support caching for multiple systems, what's the bottlenecks/other bugs i missing which results in current dora does not support full URL?

**Review by LuQQiu — COMMENTED** [review]
Thanks for the changes!
LGTM for the bug fixes e.g. file ID!
for the absolute path change to relative path part, anything in dora prevents we from keeping the absolute path or absolute path has some extra bugs?

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 12 above)
Comment (huanghua78) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:125]: The problems or fixes here are:
1. Name = A file name without SEPARATE.
2. Path = Alluxio visible path, a.k.a the path in Alluxio name space.
3. UfsPath = Full path in UFS. This ufsPath will be used later by "open()" to open the actually file as Stream.


It should be clear that: Any path/file names from Alluxio client API are Alluxio based file/path name. They are in Alluxio namespace. They must be transformed to UFS file/path name when accessing them in UFS.
This is done via Mount Tables.
On the other hand, any file/path name returned in FileStatus from UFS, must be stripped, removing the leading UFS root path.

This is the main problem here.
I don't see other bugs or bottlenecks.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java (hunk 10 above)
Comment (huanghua78) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java:180]: Let me explain the problems here.

`mFileSystem` might be DoraCacheFileSystem, or UfsBaseFileSystem.
This class is used in both UFS and Dora file system tests.
The file path in DoraCacheFileSystem is Alluxio based file path and,
the file path in UfsBaseFileSystem is absolute/full path.
Let's assume a `path` passed to this function is "/Some_Dir/Some_File".
Let's assume we are using:
alluxio.master.mount.table.root.ufs=/Users/xxx/alluxio/underFSStorage

In Dora test, the final file in host OS is:
/Users/xxx/alluxio/underFSStorage/Some_Dir/Some_File
In UfsBaseFileSystem test, the final file in host OS is:
/Some_Dir/Some_File

Before the test running, the test suite will create a file and write some data into it.
When the test is running, we will check the file.
ERROR will happen now, because in preparation, the `/Some_Dir/Some_File` is created in host OS.
But in Dora test, the expected file in host OS will be: `/Users/xxx/alluxio/underFSStorage/Some_Dir/Some_File`
This causes tests error.

That's why I am checking if the `mFileSystem` is Dora or not.
This is purely a bug/problem in test suite. Not in the Alluxio main code.

File: core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java (hunk 12 above)
Comment (LuQQiu) [reply core/server/worker/src/main/java/alluxio/worker/grpc/DoraWorkerClientServiceHandler.java:125]: 1. Name = A file name without SEPARATE. fully agree
2. Path. Alluxio name space path <-- this is the part may need discussion today
3. UfsPath <-- fully agree

If we use relative path in Alluxio, multiple UFSes will have the same relative path in Alluxio namespace, and may result in conflicts?

Since we do not have master now, can we always keep full Path, probably do not need two fields, Path == UfsPath == full ufs path

**Comment by LuQQiu** [comment]
Fully based on this PR, able to run end-to-end Fuse workloads with the bug fixes in https://github.com/LuQQiu/alluxio/commit/3e4af3cf568a4af08974b4a1a997669d770cfe83
but without changes the ufs full path to alluxio namespace relative path, this part we can discuss in the sync

**Comment by beinan** [comment]
alluxio-bot, merge this please.
